src/codeGenerator/evaluator.ts

```diff
@@ -353,14 +406,143 @@ export class Evaluator implements ExpressionVisitor {
         }
         return result;  // just return the last result
     }
-    visitOn(_astNode: On, _env: EvaluatorEnv) : EvaluatorData {
-        throw new Error("Method not implemented.");
+
+    // FIXME: lots of switch on type nonsense
+    translateItemSpec(itemSpec: ItemSpec | null, env: EvaluatorEnv, sourceExpression: Expression): Store.ItemSpec | null {
+        if (itemSpec == null) {
+            return null;
+        }
+        if (itemSpec instanceof ItemMatcher) {
+            let str = this.evaluateExpectType(itemSpec.name, env, "string", "advancement trigger item specification");
+            return new Store.ItemMatcher(str);
+        } else if (itemSpec instanceof TagMatcher) {
+            let str = this.evaluateExpectType(itemSpec.name, env, "string", "advancement trigger item specification");
+            return new Store.TagMatcher(str);
+        } else {
+            throw new DSLEvaluationError(sourceExpression, `Unknown ItemSpec class ${itemSpec.constructor.name}`);
+        }
     }
-    visitAdvancement(_astNode: Advancement, _env: EvaluatorEnv) : EvaluatorData {
-        throw new Error("Method not implemented.");
+
+    parseTrigger(trigger: Trigger, env: EvaluatorEnv, sourceExpression: Expression) : Store.Trigger[] {
+        let results: Store.Trigger[] = [];
+        if (trigger instanceof Load || trigger instanceof Tick) {
+            throw new DSLSyntaxError(sourceExpression, "load and tick triggers cannot be combined");
+        } else if (trigger instanceof ConsumeItem) {
+            let itemSpec = this.translateItemSpec(trigger.details, env, sourceExpression);
+            results.push(new Store.ConsumeItem(itemSpec));
+        } else if (trigger instanceof InventoryChanged) {
+            let itemSpec = this.translateItemSpec(trigger.details, env, sourceExpression);
+            results.push(new Store.InventoryChanged(itemSpec));
+        } else if (trigger instanceof CombinedTrigger) {
+            results = results.concat(this.parseTrigger(trigger.lhs, env, sourceExpression));
+            results = results.concat(this.parseTrigger(trigger.rhs, env, sourceExpression));
+        }
+        return results;
     }
-    visitFunction(_astNode: MCFunction, _env: EvaluatorEnv) : EvaluatorData {
-        throw new Error("Method not implemented.");
+
+    // FIXME: lots of switch on type nonsense
+    parseCommands(astCommands: Command[], env: EvaluatorEnv, sourceExpression: Expression) : string[] {
+        let commands: string[] = [];
+        for (let astCommand of astCommands) {
+            if (astCommand instanceof Grant) {
+                commands.push(`advancement grant @p only ${this.evaluateExpectType(astCommand.name, env, "string", "grant command parameter")}`);
```

Comment:
Instead of using an `if` you can switch on the type in TS like so:

```
switch(astCommand.constructor) {
   case Grant:
    ...your code here
}
```

Reply:
I'll keep this in mind when doing the refactor!